### PR TITLE
Ref #9134 - gb plugin sets configs for foreman-gutterball

### PR DESCRIPTION
**Pull request by dustints**
depends on https://github.com/Katello/puppet-gutterball/pull/6

### CODE REVIEW

File: templates/foreman_gutterball.yaml.erb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+<%# for foreman_gutterball plugin %>
+:url: <%= @fqdn %>
```
Comment (komidore64) [inline templates/foreman_gutterball.yaml.erb]: this should have `/gutterball` on the end. other than that, i will see if i can test this on monday!

**Comment by ehelms** [comment]
A more tpyical convention would be to create `manifests/plugin/gutterball/config.pp` and perhaps to mimic the same for templates to help with separation and locating things, e.g. `templates/plugin/gutterball/foreman_gutterball.yaml.erb`

File: manifests/plugin/foreman_gutterball_config.pp [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# foreman_gutterball config
+class katello::plugin::foreman_gutterball_config(
+    $foreman_user  = 'foreman',
+    $foreman_group = 'foreman',
+    $foreman_plugins_dir = '/usr/share/foreman/plugins',
```
Comment (ehelms) [inline manifests/plugin/foreman_gutterball_config.pp]: I don't think this is the right location for the file to get laid down. I believe it should be `/etc/foreman/settings.plugins.d/` . See https://github.com/theforeman/foreman/blob/develop/config/settings.rb#L14

File: templates/foreman_gutterball.yaml.erb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+<%# for foreman_gutterball plugin %>
```
Comment (ehelms) [inline templates/foreman_gutterball.yaml.erb]: I would consider namespacing your settings, e.g.:

```
:foreman_gutterball:
  :url:
```

If you don't namespace then `url` gets put into the top level settings namespace and could be over written easily. This way you can access SETTINGS[:foreman_gutterball][:url] and it's clear what 'url' is related to and les slikely to be over written.

File: manifests/plugin/foreman_gutterball_config.pp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    $foreman_gutterball_gem_config = undef
+  ){
+  $config_file = "${foreman_plugins_dir}/foreman_gutterball.yaml"
+
+
+  concat{ $config_file:
```
Comment (ehelms) [inline manifests/plugin/foreman_gutterball_config.pp]: I am not following why all this concatenation needs to occur to deploy this settings file?

File: manifests/plugin/foreman_gutterball_config.pp [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  }
+
+  concat::fragment { "foreman_gutterball.yaml from template":
+    target => $config_file,
+    content => template('katello/foreman_gutterball.yaml.erb'),
+    order => '01'
+  }
+  concat::fragment { "gem's foreman_gutterball.yaml":
+    target => $config_file,
+    source => "$foreman_gutterball_gem_config",
```
Comment (ekohl) [inline manifests/plugin/foreman_gutterball_config.pp]: Does not have to be quoted.

File: manifests/plugin/foreman_gutterball_config.pp (hunk 4 above)
Comment (dustints) [inline manifests/plugin/foreman_gutterball_config.pp]: @ehelms hoping to allow creating settings from foreman-gutterball and here.

Comment (dustints) [inline manifests/plugin/foreman_gutterball_config.pp]: @ehelms i guess concat doesn't have to happen..just a new file name would work
...will address.

File: templates/foreman_gutterball.yaml.erb (hunk 3 above)
Comment (dustints) [inline templates/foreman_gutterball.yaml.erb]: @ehelms is it ok for url to fall into something like :foreman_gutterball_common:
i think if the 2 settings use the same namespace, one will overwrite the other instead of merging them.

File: templates/foreman_gutterball.yaml.erb (hunk 1 above)
Comment (dustints) [inline templates/foreman_gutterball.yaml.erb]: addressing

**Comment by dustints** [comment]
@ehelms, @komidore64 , @ekohl 
addressed commends.
* template is now @ templates/plugin/gutterball/foreman_gutterball_addressable.yaml.erb
* does not concat with the gem's settings into one config. instead it creates a new config in /usr/foreman..settings.plugin.d/foreman_gutterball_auxiliary.yaml

File: manifests/plugin/gutterball.pp [hunk 6]
```diff
@@ -1,9 +1,26 @@
 # gutterball plugin
 class katello::plugin::gutterball{
+  $foreman_gutterball_gem_config_tmp = '/tmp/foreman_gutterball.yaml'
+
+  $package = "${foreman::plugin_prefix}gutterball"
```
Comment (ekohl) [inline manifests/plugin/gutterball.pp]: This whole block appears to be unused.

File: templates/sclpackages.erb [hunk 7]
```diff
@@ -0,0 +1 @@
+<%= @scl_package_names.map{ |pkg| @scl_prefix + pkg }.join(',') %>
```
Comment (ekohl) [inline templates/sclpackages.erb]: This appears to be unrelated.

File: manifests/plugin/gutterball/config.pp [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# gutterball config
+class katello::plugin::gutterball::config(
+    $foreman_user  = 'foreman',
+    $foreman_group = 'foreman',
+    $foreman_plugins_dir = '/usr/share/foreman/config/settings.plugins.d',
```
Comment (ekohl) [inline manifests/plugin/gutterball/config.pp]: This should be in /etc.

File: manifests/plugin/gutterball.pp [hunk 9]
```diff
@@ -1,9 +1,26 @@
...
+    }
+  }
+
   Class[ 'certs' ] ->
   class { 'certs::gutterball': } ->
   foreman::plugin { 'gutterball': } ->
   class { '::gutterball':
     keystore_password => $certs::gutterball::gutterball_keystore_password,
+  } ~>
+  class { 'katello::plugin::gutterball::config':
```
Comment (ekohl) [inline manifests/plugin/gutterball.pp]: I think this doesn't have to be a separate class since it's only a single file. I even intend to make this possible in https://github.com/theforeman/puppet-foreman/pull/281 with a parameter to ```foreman::plugin```.

File: manifests/plugin/gutterball/config.pp [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    $foreman_gutterball_gem_config = undef
```
Comment (ekohl) [inline manifests/plugin/gutterball/config.pp]: This appears to be unused.

File: manifests/plugin/gutterball.pp [hunk 11]
```diff
@@ -1,9 +1,26 @@
 # gutterball plugin
 class katello::plugin::gutterball{
+  $foreman_gutterball_gem_config_tmp = '/tmp/foreman_gutterball.yaml'
```
Comment (ekohl) [inline manifests/plugin/gutterball.pp]: This appears to be unused.

File: manifests/plugin/gutterball.pp (hunk 9 above)
Comment (dustints) [inline manifests/plugin/gutterball.pp]: that would be nice to let configs manage their own configs easily.
I kept it in it's own class so it could be reused by puppet-katello_devel https://github.com/Katello/puppet-katello_devel/pull/30

**Comment by dustints** [comment]
@ekohl  updated.

File: manifests/plugin/gutterball/config.pp [hunk 12]
```diff
@@ -0,0 +1,14 @@
+# gutterball config
+class katello::plugin::gutterball::config(
+    $foreman_user  = 'foreman',
+    $foreman_group = 'foreman',
+    $foreman_plugins_dir = '/etc/foreman/plugins',
+  ){
+
+  file { "${foreman_plugins_dir}/foreman_gutterball_auxiliary.yaml":
```
Comment (ehelms) [inline manifests/plugin/gutterball/config.pp]: Why 'auxiliary' ?

File: manifests/plugin/gutterball/config.pp [hunk 13]
```diff
@@ -0,0 +1,14 @@
+# gutterball config
+class katello::plugin::gutterball::config(
+    $foreman_user  = 'foreman',
+    $foreman_group = 'foreman',
+    $foreman_plugins_dir = '/etc/foreman/plugins',
```
Comment (ehelms) [inline manifests/plugin/gutterball/config.pp]: I still think this is the wrong directory - https://github.com/Katello/puppet-katello/pull/58#discussion-diff-23880066

File: templates/plugin/gutterball/foreman_gutterball_auxiliary.yaml.erb [hunk 14]
```diff
@@ -0,0 +1,2 @@
+:foreman_gutterball_auxiliary:
+  :url: <%= @fqdn %>:8443/gutterball
```
Comment (ekohl) [inline templates/plugin/gutterball/foreman_gutterball_auxiliary.yaml.erb]: I think this should also be parameterized.

File: manifests/plugin/gutterball/config.pp (hunk 13 above)
Comment (ekohl) [inline manifests/plugin/gutterball/config.pp]: It's not. 
```
# ls -l /usr/share/foreman/config/settings.plugins.d
lrwxrwxrwx. 1 root root 20 jan 29 18:38 /usr/share/foreman/config/settings.plugins.d -> /etc/foreman/plugins
```

File: manifests/plugin/gutterball/config.pp (hunk 12 above)
Comment (dustints) [inline manifests/plugin/gutterball/config.pp]: what's a good name...forman_gutterball_url?

Comment (ehelms) [inline manifests/plugin/gutterball/config.pp]: How about just `foreman_gutterball.yaml` ? I don't see why it needs a special name when it is simply the config file for foreman_gutterball.

**Comment by dustints** [comment]
@ehelms, @ekohl updated
* rename foreman_gutterball_aux => foreman_gutterball_url.yaml
* parameterize the url supplied in template.

File: manifests/plugin/gutterball/config.pp (hunk 12 above)
Comment (ekohl) [inline manifests/plugin/gutterball/config.pp]: I'd use gutterball.yaml since the plugin is also named gutterball. In the PR I mentioned I will also default to that if you don't specify a config filename.

File: templates/plugin/gutterball/foreman_gutterball_url.yaml.erb [hunk 15]
```diff
@@ -0,0 +1,6 @@
+:foreman_gutterball_url:
+  :protocol: <%= @protocol %>
```
Comment (ekohl) [inline templates/plugin/gutterball/foreman_gutterball_url.yaml.erb]: Does the plugin use these variables? I'd only parameterize the configuration parameters that the plugin actually uses.

Comment (dustints) [inline templates/plugin/gutterball/foreman_gutterball_url.yaml.erb]: @komidore64 do you think there is any value for foreman-gutterball to have discrete portions of the url?

File: manifests/plugin/gutterball/config.pp (hunk 13 above)
Comment (komidore64) [inline manifests/plugin/gutterball/config.pp]: @ekohl how many dirs are there for plugin configs? what is the correct one to use?

File: templates/plugin/gutterball/foreman_gutterball_url.yaml.erb (hunk 15 above)
Comment (komidore64) [inline templates/plugin/gutterball/foreman_gutterball_url.yaml.erb]: we could go either way. i was doing similar to what is found in katello's config

for example:
```yaml
common:
  # ...
  candlepin:
    url: https://localhost:8443/candlepin
```

**Comment by dustints** [comment]
@komidore64, @ehelms, @ekohl 
updated
* just using the same format as katello's config for url
* let this module manage foreman_gutterball.yaml so renaming templates and destination config file to foreman_gutterball.yaml

File: manifests/plugin/gutterball/config.pp [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+    $protocol = 'https',
+    $port = '8443',
+    $path = 'gutterball/',
```
Comment (ehelms) [inline manifests/plugin/gutterball/config.pp]: I think we typically defer to leading slashes instead of trailing slashes for URL paths.

File: manifests/plugin/gutterball/config.pp (hunk 13 above)
Comment (komidore64) [inline manifests/plugin/gutterball/config.pp]: i have had `/etc/foreman/plugins/<title>.yaml` confirmed to me on IRC.

**Comment by dustints** [comment]
@ehelms updated

**Comment by ehelms** [comment]
APT

File: templates/plugin/gutterball/foreman_gutterball.yaml.erb [hunk 17]
```diff
@@ -0,0 +1,2 @@
+:foreman_gutterball:
+  :url: <%= @protocol %>://<%= @fqdn %>:<%= @port %><%= @path %>
```
Comment (komidore64) [inline templates/plugin/gutterball/foreman_gutterball.yaml.erb]: one last thing...

let's stick with the keys being strings instead of symbols, so we can stay uniform with other configs.

so
```erb
foreman_gutterball:
  url: <%= @protocol %>://<%= @fqdn %>:<%= @port %><%= @path %>
```

**Comment by dustints** [comment]
@komidore64 updated

**Comment by komidore64** [comment]
awesome!

ack pending tests

File: manifests/plugin/gutterball/config.pp [hunk 18]
```diff
@@ -0,0 +1,17 @@
+# gutterball config
+class katello::plugin::gutterball::config(
+    $foreman_user  = 'foreman',
+    $foreman_group = 'foreman',
+    $foreman_plugins_dir = '/etc/foreman/plugins',
+    $protocol = 'https',
```
Comment (ekohl) [inline manifests/plugin/gutterball/config.pp]: I still think a single URL parameter is easier than all the separate components. Now you're still forced to use the fqdn which a user might not want.

**Comment by dustints** [comment]
@ekohl updated to a single url param

File: manifests/plugin/gutterball.pp (hunk 9 above)
Comment (ekohl) [inline manifests/plugin/gutterball.pp]: I still think you should move the file resource into this class and drop the separate ::config class because it will simplify things.

File: manifests/plugin/gutterball/config.pp [hunk 19]
```diff
@@ -0,0 +1,15 @@
+# gutterball config
+class katello::plugin::gutterball::config(
+    $foreman_user  = 'foreman',
+    $foreman_group = 'foreman',
+    $foreman_plugins_dir = '/etc/foreman/plugins',
+    $url = "https://${fqdn}:8443/gutterball",
+  ){
+
+  file { "${foreman_plugins_dir}/foreman_gutterball.yaml":
```
Comment (ekohl) [inline manifests/plugin/gutterball/config.pp]: I'd name this simply gutterball.yaml since foreman is implied.

**Comment by dustints** [comment]
@ekohl, updated renamed the soruce template & destination config file

File: manifests/plugin/gutterball/config.pp (hunk 19 above)
Comment (ehelms) [inline manifests/plugin/gutterball/config.pp]: I prefer things are explicit and match the plugin name. That way there is
no translation or discrepancy whenever you are looking for something
related. I think that the namespace of the plugins config and the file
containing it should remain the same.

Comment (ekohl) [inline manifests/plugin/gutterball/config.pp]: That's also a good point. I'll think about this and get back to you.

Comment (komidore64) [inline manifests/plugin/gutterball/config.pp]: I agree with @ehelms. the name of the plugin is `foreman_gutterball` so the plugin's config file should have the same name.

**Comment by dustints** [comment]
merging! too much time spent on this 10 liner
